### ISSUE TITLE
测试方式问题

### ISSUE THREAD

**Issue by dingfengqian** [pykt-team/pykt-toolkit#265]
感谢贵团队在领域做出的巨大贡献，我有几个问题想请教一下：

**问题 1：关于 test_sequences.csv 与 test_window_sequences.csv 的评估**

在 wandb_predict.py 中，模型会分别对 test_sequences.csv 和 test_window_sequences.csv，这两个文件调用 evaluate() 进行评估。其中 test_window_sequences.csv 是在 test_sequences.csv 的基础上，将长度超过 max_seq_len 的序列按 shift-one（滑动窗口） 拆分得到的。

想请问：
为什么需要同时对这两个测试文件进行评估？它们分别代表什么测试目的？
test_window_sequences.csv 中由 shift-one 生成的子序列高度重叠，这样在计算整体测试指标时，会不会导致某些时间步被多次计算、从而影响评估结果的解释？

**问题 2：关于测试阶段的 shift-one 单步预测方式**
在测试阶段，每个序列都会采用 shift-one 的单步预测方式，例如：

```
[1] → 预测 2  
[1,2] → 预测 3  
[1,2,3] → 预测 4  
...
```

这意味着：
序列前期可利用的历史很短，预测可能不准，序列后期历史更多，预测可能更准

因此想请问：
这种“前期信息少、后期信息多”的差异，是否会影响整体测试指标的公平性？

**Comment by lycyhrc** [comment]
感谢您的关注！

**问题1：关于 test_sequences.csv 与 test_window_sequences.csv 的评估**
这两个文件代表不同的序列切分策略，用于评估模型在**不同场景**下的性能
- test_sequences.csv（非重叠切分)：采用非重叠的方式切分长序列，每个窗口之间互不重叠，末尾长度小于 min_seq_len 的部分用-1补到长度200
- test_window_sequences.csv（滑动窗口切分）：，采用shift-one（每次滑动1步）的方式切分，窗口之间高度重叠（重叠度为 maxlen-1），不丢失任何交互数据，提供更完整的评估覆盖

> 滑动窗口不会导致重复计算，代码通过 selectmasks 机制避免了重复计算问题。

关键代码在 split_datasets.py：
`dres["selectmasks"].append(
    ",".join([str(pad_val)] * (maxlen - 1) + ["1"]))`

这段代码表明：对于每个滑动窗口，只有最后一个位置（新增的时间步）的 selectmask 被设置为 1，前面 maxlen-1 个位置（与前一个窗口重叠的部分）的 selectmask 都被设置为 -1（pad_val）。

在评估时（evaluate_model.py）：
`y = torch.masked_select(y, sm).detach().cpu()  # 只选择 sm=1 的预测`
`t = torch.masked_select(rshft, sm).detach().cpu()  # 只选择 sm=1 的真实标签`

每个时间步只会被计算一次 —— 即在它作为窗口最后一个位置时被计算。


总的来说，window_testauc 和 window_testacc 通常更可靠，因为它们评估了所有时间步。两个指标都有参考价值，若差异较大，可能说明模型对序列边界或上下文长度敏感，但是大多数kt数据集差异不大。

**问题 2：历史长度差异是否影响公平性？** 

是的，确实会影响。我们曾经做过不同序列长度的实验：
序列前期（历史少）→ 预测通常更不准确，序列后期（历史多）→ 预测通常更准确。 

pykt当前的整体指标（AUC/ACC）是所有位置的平均，包含了冷启动短序列场景，但没有考虑不同长度的差异。
如果你想更细致地分析，可以按交互长度分层评估，主要有两个思路：

1. 学生层面：选一些长度不等的学生单独预测，短历史（< 10）、中等历史（10-50）、长历史（> 50）来计算 AUC
2. 交互层面：通过改变数据预处理中的 maxlen（如 20、50、200）观察模型表现的变化

**Comment by dingfengqian** [comment]
感谢解答
